TTS/tts/utils/text/__init__.py

```diff
@@ -25,6 +26,33 @@
 # Regular expression matching punctuations, ignoring empty space
 PHONEME_PUNCTUATION_PATTERN = r"[" + _punctuations.replace(" ", "") + "]+"
 
+# language -> source phoneme -> dest phoneme
```

Comment:
Why do we need this additional mapping?

Reply:
gruut and eSpeak don't agree completely on the phoneme inventories for each language. Some differences are cosmetic, like the use of `/ɡ/` (0x261) vs. `/g/` (0x67) -- both are acceptable IPA.

I made choices with some languages that may cause problems in multi-lang models. For example, my Dutch phonemes list uses `/ɹ/` and `/w/` when it should really be using `/r/` and `/ʋ/` (I know better now). These are easy to fix with a static map, and I plan to correct them in later versions of gruut.

Other differences are more problematic, and I'm not sure exactly what to do. Consider the sentence "responsible bee city". eSpeak phonemizes it like this:

```sh
$ espeak-ng -v en-us -qx --ipa --sep=' ' 'responsible bee city'
 ɹ ᵻ s p ˈɑː n s ᵻ b əl  b ˈiː  s ˈɪ ɾ i
```

Each "e" sound is a little different: `/ᵻ/`, `/iː/`, or `/i/`

gruut is much more simplistic:

```sh
$ bin/gruut en tokenize 'responsible bee city' | bin/gruut en phonemize | jq -r .pronunciation_text
ɹ i s p ˈɑ n s ɪ b ə l b ˈi s ˈɪ t i
```

Now they're all just `/i/` because gruut was designed to operate on **phonemes** rather than phones. I assumed that the realization of the phoneme (short, long, etc.) would be learned by the machine learning model (which seems to have worked at least in Larynx).

Some options:

* Fix cosmetic/multi-lang phonemes with a built-in static phoneme map
* Allow existing eSpeak-based models to include a custom phoneme map in their configs
* Re-train available models with gruut phoneme sets

Reply:
@synesthesiam : I don't understand why you say `gruut is much more simplistic` for the sentence `responsible bee city` ? The generated phonemes are defined this way in your `/en-us/lexicon.txt` :

```    
bee _ b ˈi
bee's _ b ˈi z

city _ s ˈɪ t i
city's _ s ˈɪ t i z

responsibility _ ɹ i s p ˌɑ n s ɪ b ˈɪ l ə t i
responsible _ ɹ i s p ˈɑ n s ɪ b ə l
responsibly _ ɹ ɪ s p ˈɑ n s ɪ b l i
```   

If you change the pronunciation in the english lexicon, create a new SQlite lexicon.db and a new g2p.fst model, you get the same results as espeak-ng.

Here is the result for my luxembourgish gruut model with the luxembourgish **The North Wind and the Sun** sentence : 
```   
mbarnig@mbarnig-MS-7B22:~/gruut$ bin/gruut lb tokenize 'An der Zäit hunn sech den Nordwand an d’Sonn gestridden, wie vun hinnen zwee wuel méi staark wier, wéi e Wanderer, deen an ee waarme Mantel agepak war, iwwert de Wee koum. ' | bin/gruut lb phonemize | jq -r .pronunciation_text

ɑ n d ɐ ts æːɪ t h u n z ə ɕ d ə n n ɔ ʀ d v ɑ n t ɑ n  g ə ʃ t ʀ i d ə n | v iə f u n h i n ə n ts w eː v uə l m ɜɪ ʃ t aː ʀ k v iː ɐ | v ɜɪ ə v ɑ n d ə ʀ ɐ | d eː n ɑ n eː v aː ʀ m ə m ɑ n t ə l  v aː ʀ | i v ɐ t d ə v eː k əʊ m ‖
```

Reply:
@mbarnig, you're correct -- this is not a technical limitation of gruut itself. It's just that the phoneme set I selected for U.S. English is smaller than what eSpeak has.

This has got me thinking of a new idea,  though. Perhaps I should store pronunciations in their original (larger) phoneme set, and then allow the user to select their desired phoneme set. This would solve one problem I've had with needing to convert between various ASR phoneme sets.

Another facet of this idea that I'd like to explore more is "accented speech", where you approximate the phoneme set of one language in another. I have a toy version of this already in gruut/larynx using manually created maps between a few languages. It would be cool to see if it could be partially automated, so you could for example use an English TTS voice to speak a French sentence.

Reply:
@synesthesiam I agree that using simpler phoneme representation would be adjusted by the model. So I am fine with using simple `/i/` in that example. Also if we go a level deeper, then we need to generate phonemes contextually and it is a really hard task.

What I don't understand is that IPA representations should be universal and if it is the case (I might be wrong), why do we need an explicit static mapping for the languages? Is it just for compatibility with your previous model or am I missing here? 

From an ML perspective, what is important for 🐸TTS is that the phonemizer, in this case gruut, should generate the same sequence of phonemes for the same sequence of sounds. This is actually what needs to happen in G2 for any language,  right? That's what I mean by saying IPA is universal.

Reply:
@synesthesiam : I tested your idea with the luxembourgish sentence 
`An der Zäit hunn sech den Nordwand an d’Sonn gestridden.`    
The correct IPA phonemization is
`|ɑ|n| |d|ɐ| |ʦ|æ:ɪ|t| |h|u|n| |z|e|ɕ| |d|ə|n| |n|o|ʀ|t|v|ɑ|n|t| |ɑ|n| |d|z|o|n |g|ə|ʃ|t|ʀ|i|d|ə|n|.`  
The german espeak-ng phonemization for this sentence is   
`|a|n| |d|ɛ|ɾ| |ts|ɛː|ˈɪ|t| |h|ˈʊ|n| |z|ˈɛ|ç| |d|eː|n |n|ˈɔ|ɾ|d|v|a|n|t| |a|n| |d|ˈeː|z|ˈɔ|n| |ɡ|ə|ʃ|t|ɾ|ˈɪ|d|ə|n|.`  
The synthesized speech with the german model is not really intelligible. 
I changed three phonemes in the luxembourgish phoneme sequence, as follows :
`|ɑ|n| |d|ɐ| |ts|aɪ|t| |h|u|n| |z|e|ç| |d|ə|n| |n|o|ʀ|t|v|ɑ|n|t| |ɑ|n| |d|z|o|n |g|ə|ʃ|t|ʀ|i|d|ə|n|.`  
Now Thorsten (@thorstenMueller) speaks luxembourgish, with a german accent.

![coqui-tts-server-lb](https://user-images.githubusercontent.com/1360633/120928005-3db3c700-c6e3-11eb-9301-e92af8305a2f.png)

@erogol ; @synesthesiam : I think we have a different understanding and experience of IPA. Perhaps we should investigate in depth how to build the multilingual phonemizer to get a clear view.

Reply:
@erogol I agree that the most important aspect is that the same phonemes are produced in the same context.

I'm not an expert in IPA, but I would say it's "universal" in the sense that it's a formal notation for how humans vocalize. But the level of detail varies wildly, spanning from perfectly capturing how one person said a specific utterance up to high-level phonemes that all speakers of a language share. Depending on the task, you'll likely choose a different level of detail.

Compared to espeak, gruut currently has "lower resolution" phonemes. That matters when trying to use gruut with pre-trained :frog: TTS models that were using espeak's phones/phonemes, but my question is: does :frog: TTS need the "higher resolution" or will newly-trained models simply figure it out themselves (like they do for intonation)?

Reply:
@synesthesiam yep my background is also ML thus I know phonemes as much as I know them from dictionaries :) 

Therefore, I believe it is better to delegate all these complexities to the model and the dataset to allow more people to train models. 

So far,  I see that if the dataset is good then the model works with rough phonemes and even sometimes it adjusts some phoneme errors. 

In the end, for the model, it is important to see the same sequence of phoneme characters for the same sound consistently. It does not care if the phonemization is correct or not.

However, the downside is that then the model would not produce consistent results for raw phoneme inputs.  

But here the famous 80-20 rule applies I guess. I guess first we need to have a consistent G2P interface that is easy to use. Then let people try and refine the process by collecting feedback. The feedback part is also important since different languages also have different needs. 

It is probably not the right place but I post it anyways since we are all here. I think the imperatives of G2P in 🐸TTS right now are

1. Text normalization (numbers, acronyms, titles, etc.)
2. Consistent G2P
3. Easy to use so that people with no or little knowledge can still train models.  
4. Cover as many langs as possible.

Would you agree with these?

Reply:
@erogol I would definitely agree with those 4 points.

> However, the downside is that then the model would not produce consistent results for raw phoneme inputs.

gruut might be able to help a bit here. If you take a look at the [phonemes.txt file for U.S. English](https://github.com/rhasspy/gruut-ipa/blob/master/gruut_ipa/data/en-us/phonemes.txt), the values after each example word are alternative forms for the phoneme. gruut is capable of [splitting/normalizing a raw phoneme string](https://github.com/rhasspy/gruut-ipa/blob/480e284d6c9602f9f0025559d9528de9ba6353ad/gruut_ipa/__init__.py#L811), dipthongs and all. 

By the way, my background is not in ML (nor even computational linguistics), so please feel free to correct me. In fact, a good deal of my ML knowledge has come from hacking on your code :slightly_smiling_face:

Reply:
> Now Thorsten (@thorstenMueller) speaks luxembourgish, with a german accent.

This is great, @mbarnig! I would think this technique will work quite well for related language families. I also expect voices trained on phoneme-rich languages to be more versatile.

Knowing exactly how to alter the mismatched phonemes is the hard part (for me), so finding multi-lingual people is key!